### PR TITLE
Multiple public IP support for Azure Firewall

### PR DESCRIPTION
**Pull request by iatodoro**
## Description
* Add support for multiple public IP addresses for Azure Firewall
    - Updated New-AzFirewall cmdlet:
        - Added parameter -PublicIpAddress which accepts one or more Public IP Address objects
        - Added parameter -VirtualNetwork which accepts a Virtual Network object
        - Added methods AddPublicIpAddress and RemovePublicIpAddress on firewall object - these accept a Public IP Address object as input
        - Deprecated parameters -PublicIpName and -VirtualNetworkName 
        - Added tests for the new parameters
        - Updated documentation to show how to use the new parameters and how to assign additional Public IP addresses to the firewall

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

**Comment by cormacpayne** [comment]
@iatodoro should these changes target the `network-april` branch?

CC: @MikhailTryakhov

**Comment by niander** [comment]
@iatodoro Do we have a cmdlet design review for the public API changes? Please link it in here.

**Comment by iatodoro** [comment]
> @iatodoro Do we have a cmdlet design review for the public API changes? Please link it in here.

Is a design review necessary? We're not removing any parameters or adding new cmdlets.

**Comment by iatodoro** [comment]
> @iatodoro should these changes target the `network-april` branch?

Yes, sorry about that. I rebased on top of `network-april` now.

**Comment by iatodoro** [comment]
> @iatodoro Do we have a cmdlet design review for the public API changes? Please link it in here.

https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/311

**Comment by iatodoro** [comment]
Can I get some attention on this PR? The tests seem to be failing but are passing fine on my local dev environment. We need this to make the next release cutoff.

**Comment by iatodoro** [comment]
Latest commit addresses design review comments.

File: src/Network/Network/AzureFirewall/NewAzureFirewallCommand.cs [hunk 1]
```diff
@@ -54,22 +55,56 @@ public class NewAzureFirewallCommand : AzureFirewallBaseCmdlet
         [ValidateNotNullOrEmpty]
         public virtual string Location { get; set; }
 
+        [CmdletParameterBreakingChange(
+            "VirtualNetworkName",
+            deprecateByVersion: "2.0.0",
+            ChangeDescription = "This parameter will be removed in an upcoming breaking change release. After this point the Virtual Network will be provided as an object instead of a string.", 
+            OldWay = "New-AzureRmFirewall -VirtualNetworkName \"vnet-name\"",
+            NewWay = "New-AzureRmFirewall -VirtualNetwork $vnet",
```
Comment (MiYanni) [inline src/Network/Network/AzureFirewall/NewAzureFirewallCommand.cs]: `OldWay` and `NewWay` should call `New-AzFirewall`, not AzureRM.

File: src/Network/Network/AzureFirewall/NewAzureFirewallCommand.cs [hunk 2]
```diff
@@ -54,22 +55,56 @@ public class NewAzureFirewallCommand : AzureFirewallBaseCmdlet
...
             HelpMessage = "Virtual Network Name")]
         [ValidateNotNullOrEmpty]
         public string VirtualNetworkName { get; set; }
 
+        [CmdletParameterBreakingChange(
+            "PublicIpName",
+            deprecateByVersion: "2.0.0",
+            ChangeDescription = "This parameter will be removed in an upcoming breaking change release. After this point the Public IP Address will be provided as a list of one or more objects instead of a string.",
+            OldWay = "New-AzureRmFirewall -PublicIpName \"public-ip-name\"",
+            NewWay = "New-AzureRmFirewall -PublicIpAddress @($publicip1, $publicip2)",
```
Comment (MiYanni) [inline src/Network/Network/AzureFirewall/NewAzureFirewallCommand.cs]: `OldWay` and `NewWay` should call `New-AzFirewall`, not AzureRM.

File: src/Network/Network/help/New-AzFirewall.md [hunk 3]
```diff
@@ -359,6 +414,10 @@ This cmdlet supports the common parameters: -Debug, -ErrorAction, -ErrorVariable
 
 ### System.String
 
+### Microsoft.Azure.Commands.Network.Models.PSVirtualNetwork
+
+### System.Collections.Generic.List`1[[Microsoft.Azure.Commands.Network.Models.PSPublicIpAddress, Microsoft.Azure.PowerShell.Cmdlets.Network, Version=1.8.1.0, Culture=neutral, PublicKeyToken=null]]
```
Comment (MiYanni) [inline src/Network/Network/help/New-AzFirewall.md]: It looks like this wasn't updated when the parameter was changed from a List to and array.

File: src/Network/Network/AzureFirewall/NewAzureFirewallCommand.cs (hunk 1 above)
Comment (iatodoro) [reply src/Network/Network/AzureFirewall/NewAzureFirewallCommand.cs]: Done, thank you.

File: src/Network/Network/help/New-AzFirewall.md (hunk 3 above)
Comment (iatodoro) [reply src/Network/Network/help/New-AzFirewall.md]: Done, thank you.

**Comment by number213** [comment]
Test on Linux keeps failing because of DataLake's test

```
Starting test execution, please wait...
[xUnit.net 00:00:26.73]     Microsoft.Azure.Commands.DataLakeStore.Test.ScenarioTests.AdlsAliasTests.TestAdlsFileSystem [FAIL]
Failed   Microsoft.Azure.Commands.DataLakeStore.Test.ScenarioTests.AdlsAliasTests.TestAdlsFileSystem
EXEC : error Message:  [/home/vsts/work/1/s/build.proj]
System.Management.Automation.ActionPreferenceStopException : The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: Object reference not set to an instance of an object.
```

The worst part is that whenever an error occurs Xunit sends all test output to stdout. In TestAdlsFileSystem's case that's more than 2 million lines of text. This stalls test job until 2 hour timeout is reached.
@MiYanni could you please take a look into that?
Example CI jobs:
- https://dev.azure.com/azure-sdk/public/_build/results?buildId=43069
- https://dev.azure.com/azure-sdk/public/_build/results?buildId=43226

**Comment by number213** [comment]
@MiYanni @cormacpayne could you please merge this PR if everything looks good?

**Pull request merged**